mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/definition/annotation/FullTextFilter.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.pojo.mapping.definition.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.processing.TypeMapping;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.processing.TypeMappingAnnotationProcessorRef;
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.processing.impl.FullTextFilterProcessor;
+import org.hibernate.search.mapper.pojo.bridge.mapping.annotation.FilterFactoryRef;
+
+/**
+ * Maps an indexed type to its filter using a {@link FullTextFilter},
+ * <p>
+ * See the reference documentation for more information about bridges in general,
+ * and filter in particular.
+ */
+@Documented
+@Target({ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Repeatable(FullTextFilter.List.class)
+@TypeMapping(processor = @TypeMappingAnnotationProcessorRef(type = FullTextFilterProcessor.class))
+public @interface FullTextFilter {
+
+	/**
+	 * @return The name of the filter.
+	 */
+	String name();
+
+	/**
+	 * @return A reference to the factory filter to use.
+	 * @see FilterFactoryRef
+	 */
+	FilterFactoryRef factory();
+
+	/**
+	 * @return A list of components in the paths, each representing a property and a way to extract its value.
+	 */
+	FilterParam[] params() default {};
```

Comment:
The idea was to shorten the solution to the definition of a factory. 
Currently, it can be defined differently: 
```java
public class IsActiveBinder  implements TypeBinder {
    @Override
    public void bind(TypeBindingContext context) {
        context.indexSchemaElement().namedPredicate("is-active", new PricesActiveFilterFactory());
    }

    public class PricesActiveFilterFactory implements NamedPredicateFactory {

        @Override
        public SearchPredicate create(NamedPredicateFactoryContext ctx) {
            SearchPredicate filter;
            String fieldPath = ctx.resolvePath("active");

            filter = ctx.predicate()
                    .match().field(fieldPath)
                    .matching(true)
                    .toPredicate();

            return filter;
        }
    }
}
```

```java
@TypeBinding(binder = @TypeBinderRef(type = IsActiveBinder.class))
public class AssortmentPrice implements Serializable {
........
```
The annotation can also be done differently in this case: 

```java
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.TYPE})
@TypeMapping(processor = @TypeMappingAnnotationProcessorRef(
        type = IsActiveBinding.Processor.class
))
@Documented
public @interface IsActiveBinding {

    String name() default "is-active";

    class Processor implements TypeMappingAnnotationProcessor<IsActiveBinding> {

        @Override
        public void process(TypeMappingStep mapping, IsActiveBinding annotation,
                TypeMappingAnnotationProcessorContext context) {

            IsActiveBinder binder = new IsActiveBinder(annotation.name());
            mapping.binder(binder);
        }
    }
}
```

Maybe remove the annotations completely? And leave the definition of "named predicate" in TypeBinding. 

Or simplify the annotation as making it easier to define named predicate.

Reply:
But building the "TypeMappingStep.namedPredicate (String name, NamedPredicateBinder binder)" function does not seem to make sense anymore, because you refer to "TypeMappingStep.binder (TypeBinder binder)", using it you can achieve the same goal. I will remove the "" TypeMappingStep.namedPredicate "and" NamedPredicateBinder "functions completely"

Reply:
In virtually every solution, the code used to accomplish this task made no sense. I'll delete everything in "org.hibernate.search.mapper.pojo". This can be done more easily, even if the reference to "NamedPredicateFactory" was to be shortened. 

Użycie "NamedPredicateBinderRef binder", takze nie ma sensu skoro można było by to samo zrealizować za pomocą "
```java
@TypeBinding(binder = @TypeBinderRef(type = IsActiveBinder.class))
```

Reply:
Ale można rozważyć proponowaną funkcję skracającą definicję "NamedPredicateFactory", w nowej wersji zawiera nei wiele kodu:
```java
@NamedPredicateBinding(name="is-active", factory = @NamedPredicateFactoryRef(type = PricesActiveFilterFactory.class))
```
To do this, a simple "TypeMappingAnnotationProcessor" is enough: 
```java
public final class NamedPredicateBindingProcessor implements TypeMappingAnnotationProcessor<NamedPredicateBinding> {

	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );

	@Override
	public void process(TypeMappingStep mapping, NamedPredicateBinding annotation,
			TypeMappingAnnotationProcessorContext context) {
		String name = annotation.name();

		Map<String, Object> filterParams = new LinkedHashMap<>();
		NamedPredicateParam[] params = annotation.params();
		if ( params != null ) {
			for ( NamedPredicateParam param : params ) {
				filterParams.put( param.name(), param.value() );
			}
		}

		BeanReference<? extends NamedPredicateFactory> factoryReference = createFactoryReference( annotation.factory(), context );

		TypeBinder binder = new NamedPredicateTypeBinder( name, factoryReference, filterParams );
		mapping.binder( binder );
	}

	private BeanReference<? extends NamedPredicateFactory> createFactoryReference(NamedPredicateFactoryRef factoryReferenceAnnotation,
			MappingAnnotationProcessorContext context) {

		Optional<BeanReference<? extends NamedPredicateFactory>> factoryReference = context.toBeanReference(
				NamedPredicateFactory.class,
				NamedPredicateFactoryRef.PojoUndefinedFactoryImplementationType.class,
				factoryReferenceAnnotation.type(), factoryReferenceAnnotation.name(),
				factoryReferenceAnnotation.retrieval()
		);

		if ( !factoryReference.isPresent() ) {
			throw log.missingFactoryReferenceInBinding();
		}

		return factoryReference.get();
	}

	private static class NamedPredicateTypeBinder implements TypeBinder {

		private final String name;
		private final BeanReference<? extends NamedPredicateFactory> factoryReference;
		private final Map<String, Object> params;

		public NamedPredicateTypeBinder(String name, BeanReference<? extends NamedPredicateFactory> factoryReference, Map<String, Object> params) {
			this.name = name;
			this.factoryReference = factoryReference;
			this.params = params;
		}

		@Override
		public void bind(TypeBindingContext context) {
			NamedPredicateFactory factory = factoryReference.resolve( context.beanResolver() ).get();
			context.indexSchemaElement().namedPredicate( name, factory ).params( params );
		}
	}
}
```

Reply:
Maybe rename the annotation to "Name Predicate"? Keep the current functionality. As faster loading of "NamedPredicateFactory", without the end developer needing to use Binding components? 

```java
@NamedPredicate(name="is-active", factory = @NamedPredicateFactoryRef(type = PricesActiveFilterFactory.class))
public class AssortmentPrice implements Serializable {
...........
```

Reply:
I have removed the code duplicating the TypeBinding function.

If the same could be achieved by implementing an already existing TypeBinder, this redundancy of code made no sense. The developer has the ability to bus complex NamedPredicate using the functionality provided by TypeBinding. 

However, for a simple solution in many cases the previous solution may be useful. Of course, everyone has the opportunity to build their own function that allows you to bind simple factories for named predicate. But you can hide all the binding code in simple cases by providing a reference shortcut directly to the factory. I propose to name such shortcuts simply without the Binding ending: "@NamedPredicate" with the reference "NamedPredicateFactoryRef" to "NamedPredicateFactory". The more that it does not require a lot of code and interference in the entire system. The shortcut code would only consist of: 

```java
@Incubating
@Documented
@Target({ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
@Repeatable(NamedPredicate.List.class)
@TypeMapping(processor = @TypeMappingAnnotationProcessorRef(type = NamedPredicateProcessor.class))
public @interface NamedPredicate {

	/**
	 * @return The name of the named predicate.
	 */
	String name();

	/**
	 * @return A reference to the factory named predicate to use.
	 * @see NamedPredicateFactoryRef
	 */
	NamedPredicateFactoryRef factory();

	/**
	 * @return A list of components in the paths, each representing a property and a way to extract its value.
	 */
	NamedPredicateParam[] params() default {};

	@Documented
	@Target({ElementType.TYPE})
	@Retention(RetentionPolicy.RUNTIME)
	@interface List {
		NamedPredicate[] value();
	}
}
```
```java
public class NamedPredicateTypeBinder implements TypeBinder {

	private final String name;
	private final BeanReference<? extends NamedPredicateFactory> factoryReference;
	private final Map<String, Object> params;

	public NamedPredicateTypeBinder(String name, BeanReference<? extends NamedPredicateFactory> factoryReference, Map<String, Object> params) {
		this.name = name;
		this.factoryReference = factoryReference;
		this.params = params;
	}

	@Override
	public void bind(TypeBindingContext context) {
		context.dependencies().useRootOnly();
		NamedPredicateFactory factory = factoryReference.resolve( context.beanResolver() ).get();
		context.indexSchemaElement().namedPredicate( name, factory ).params( params );

		context.bridge( (target, bridgedElement, bcontext) -> {
		} );
	}
}
```
```java
public final class NamedPredicateProcessor implements TypeMappingAnnotationProcessor<NamedPredicate> {

	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );

	@Override
	public void process(TypeMappingStep mapping, NamedPredicate annotation,
			TypeMappingAnnotationProcessorContext context) {
		String name = annotation.name();

		Map<String, Object> filterParams = new LinkedHashMap<>();
		NamedPredicateParam[] params = annotation.params();
		if ( params != null ) {
			for ( NamedPredicateParam param : params ) {
				filterParams.put( param.name(), param.value() );
			}
		}

		BeanReference<? extends NamedPredicateFactory> factoryReference = createFactoryReference( annotation.factory(), context );

		TypeBinder binder = new NamedPredicateTypeBinder( name, factoryReference, filterParams );
		mapping.binder( binder );
	}

	private BeanReference<? extends NamedPredicateFactory> createFactoryReference(NamedPredicateFactoryRef factoryReferenceAnnotation,
			MappingAnnotationProcessorContext context) {

		Optional<BeanReference<? extends NamedPredicateFactory>> factoryReference = context.toBeanReference(
				NamedPredicateFactory.class,
				NamedPredicateFactoryRef.PojoUndefinedFactoryImplementationType.class,
				factoryReferenceAnnotation.type(), factoryReferenceAnnotation.name(),
				factoryReferenceAnnotation.retrieval()
		);

		if ( !factoryReference.isPresent() ) {
			throw log.missingFactoryReferenceInBinding();
		}

		return factoryReference.get();
	}
}
```
The solution is not necessary, but it will make your work much easier by not forcing you to build a complicated TypeBinding for simple named predicate use cases, e.g. 

```java
@NamedPredicate(name="is-readable", factory = @NamedPredicateFactoryRef(type = TestAccesFactory.class), params = {
    @NamedPredicateParam(name = "mode", value = "read")
})
@NamedPredicate(name="is-writable", factory = @NamedPredicateFactoryRef(type = TestAccesFactory.class), params = {
    @NamedPredicateParam(name = "mode", value = "write")
})
public class Book {
    
    @Column
    @GenericField
    private String accessMode;
    
}
```

But I leave it for you to consider adding "@NamedPredicate" as a shortcut to quickly referencing named predicate factories. I will leave the new solution for now, it can always be easily removed.

Reply:
I agree it can make things easier in some cases, but it also forces us to expose the `params( ... )` call after `.namedPredicate( name, factory )`, which is not consistent with how parameters are usually passed to the pluggable components (usually it's through a custom binder).

Also, I'd argue that once we solve [HSEARCH-3608](https://hibernate.atlassian.net/browse/HSEARCH-3608), developers will be able to solve this problem quite easily without the need of `@NamedPredicate`:

```java
@TypeBinding(binder = @TypeBinderRef(type = TestAccessFactory.Binder.class),
    params = {
        @Param(name = "name", value = "is-readable"),
        @Param(name = "mode", value = "read")
    })
@TypeBinding(binder = @TypeBinderRef(type = TestAccessFactory.Binder.class),
    params = {
        @Param(name = "name", value = "is-writable"),
        @Param(name = "mode", value = "write")
    })
public class Book {
    
    @Column
    @GenericField
    private String accessMode;
    
}
```

With `TestAccessFactory` being defined as:

```java
public class TestAccessFactory implements NamedPredicateFactory {

    // ... omitting pre-existing code ...

    // This is the only additional code you need compared to `@NamedPredicate`:
    public static class Binder implements TypeBinder {
        @Override
        public void bind(TypeBindingContext context) {
            context.namedPredicate(
                    context.param("name", String.class),
                    new TestAccessFactory(context.param("mode", String.class))
            );
        }
    }
}
```

Let's remove `@NamedPredicate` and all related code and discuss this again in another ticket if you really need it in Hibernate Search? We can work on improving usability in [HSEARCH-3608](https://hibernate.atlassian.net/browse/HSEARCH-3608).

Reply:
You can actually reduce the code a lot. I'll change it.

Reply:
To be clear I wasn't suggesting to address HSEARCH-3608 and parameter in this PR. Please don't do that in this PR! Just remove `@NamedPredicate`, we'll address HSEARCH-3608 later.